Classes/Domain/Index/Queue/QueueItemRepository.php

```diff
@@ -0,0 +1,641 @@
+        // determine which items to index with this run
+        $indexQueueItemRecords = $queryBuilder
+            ->select('*')
+            ->from($this->table)
+            ->andWhere(
+                $queryBuilder->expr()->eq('root', $site->getRootPageId()),
+                $queryBuilder->expr()->gt('changed', 'indexed'),
+                $queryBuilder->expr()->lte('changed', time()),
+                $queryBuilder->expr()->eq('errors', $queryBuilder->createNamedParameter(''))
+            )
+            ->orderBy('indexing_priority', 'DESC')
+            ->addOrderBy('changed', 'DESC')
+            ->addOrderBy('uid', 'DESC')
+            ->setMaxResults($limit)
+            ->execute()->fetchAll();
+
+        return $this->getIndexQueueItemObjectsFromRecords($indexQueueItemRecords);
+    }
+
+    /**
+     * Creates an array of ApacheSolrForTypo3\Solr\IndexQueue\Item objects from an array of
+     * index queue records.
+     *
+     * @param array $indexQueueItemRecords Array of plain index queue records
+     * @return array Array of ApacheSolrForTypo3\Solr\IndexQueue\Item objects
+     */
+    protected function getIndexQueueItemObjectsFromRecords(array $indexQueueItemRecords) : array
+    {
+        $indexQueueItems = [];
+        $tableUids = [];
+        $tableRecords = [];
+
+        // grouping records by table
+        foreach ($indexQueueItemRecords as $indexQueueItemRecord) {
+            $tableUids[$indexQueueItemRecord['item_type']][] = $indexQueueItemRecord['item_uid'];
+        }
+
+        // fetching records by table, saves us a lot of single queries
+        foreach ($tableUids as $table => $uids) {
+            $uidList = implode(',', $uids);
+
+            $queryBuilderForRecordTable = $this->getQueryBuilder();
+            $queryBuilderForRecordTable->getRestrictions()->removeAll();
+            $resultsFromRecordTable = $queryBuilderForRecordTable
+                ->select('*')
+                ->from($table)
+                ->where($queryBuilderForRecordTable->expr()->in('uid', $uidList))
+                ->execute();
+            $records = [];
+            while ($record = $resultsFromRecordTable->fetch()) {
+                $records[$record['uid']] = $record;
+            }
+
+            $tableRecords[$table] = $records;
+
+            if (is_array($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['solr']['postProcessFetchRecordsForIndexQueueItem'])) {
+                $params = ['table' => $table, 'uids' => $uids, 'tableRecords' => &$tableRecords];
+                foreach ($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['solr']['postProcessFetchRecordsForIndexQueueItem'] as $reference) {
+                    GeneralUtility::callUserFunction($reference, $params, $this);
+                }
+                unset($params);
+            }
+        }
+
+        // creating index queue item objects and assigning / mapping
+        // records to index queue items
+        foreach ($indexQueueItemRecords as $indexQueueItemRecord) {
```

Comment:
I would propose to extract two additional protected methods from line 546-570 getAllQueueItemRecordsByUidsGroupedByTable and 574-592 getQueueItemObjectsByRecords